### ISSUE TITLE
unifi-poller+unifi: Add support for UXG

### ISSUE THREAD

**Issue by light-industrial** [unifi-poller/unifi-poller#296]
Hi,
I have an early access [UXG pro](https://evanmccann.net/blog/unifi-uxg-pro), I have unifipoller working great (thank you!!) for all dashboards except the USG dashboard.  Am guessing poller might need some updates to support the new gateway hardware?  From what i read it's more similar to UDM.  I'm happy to provide debug info or help out as I can to enable support.

Thanks David & UnifiPoller!

**Comment by davidnewhall** [comment]
Can you describe the problem?

**Comment by light-industrial** [comment]
Sure, I get no data in any of the panels for the USG dashboard.  All the other unifi-poller dashboards work fine.  I assumed it was because my unifi gateway is an early access UXG Pro, should unifi-poller already support the UXG Pro?

**Comment by davidnewhall** [comment]
No clue, never used one. :) Does it show up as a device in the menu at the top?

**Comment by light-industrial** [comment]
hehe, I just noticed that the Site doesn't show up for the USG dashboard.  It does for the other dashboards.   Looking at the logs:

unifi-poller_1  | 2021/01/11 20:15:16.293943 devices.go:47: [DEBUG] Unmarshalling Device Type: uxg, site Cloud (default) 
unifi-poller_1  | 2021/01/11 20:15:16.293965 devices.go:76: [ERROR] unknown asset type - uxg - skipping

I guess this could be related? :)

**Comment by davidnewhall** [comment]
Yup. That's the kinda data I need. Please run this command send me the output. Either on here, in a gist, via email or on discord.
```
unifi-poller -j d -c /path/to/up.conf
```
 Having that output will allow me to add this new device. Thank you!

**Comment by light-industrial** [comment]
I have the output for you though need to share privately given contents.  Discord is throwing me an error, is there an email address I can send to?

**Comment by davidnewhall** [comment]
david at sleepers dot pro

**Comment by davidnewhall** [comment]
Thanks for the json dump! The UXG structure looks a lot like the UDM, so this shouldn't be too hard to work in.

**Comment by light-industrial** [comment]
Great, thanks David.  I just noticed that the client DPI dashboard is also empty, which makes sense given that DPI comes from the gateway.  Will the UXG dump be sufficient to fix that also?

Finally, I'm using the docker-compose package for Unifi-poller, once the new version is ready will it take some time for the changes to be merged into that?  Ala, should I drop docker-compose and move to actual installs of Influx, Grafana and Unifi-poller or, can I leave it as is?

**Comment by absoluteg** [comment]
I'm in the same boat, have UXG Pro and did the docker-compose method.  My docker running just outputs unknown device UXG, and I get no data from the controller.  I did setup to pull from unifi-poller master (instead of latest) to see if that made any difference, it hasn't.  Not sure what if anything I can do for now, I think just wait? I appreciate any work that goes into this.  Just to give a little more detail as the UXG Pro puts us in a beta train with unifi, I'm running cloud key gen2 + on 2.0.27 firmware. UXG Pro was updates recently as well 0.5 > 1.9.  I'm running 6.1.61 Network and 1.18.1-beta Protect.

**Comment by davidnewhall** [comment]
I have not yet added UXG support. It may be a 1 line change, it may be more. I just haven't had time yet.

**Comment by absoluteg** [comment]
Thanks, if you need any data or log outputs let me know.  More than happy
to help.

Thanks,
Greg

**Comment by light-industrial** [comment]
Hi David,
if unifi-poller already supports the UDM, then there is a pretty good
chance that it is just a one line change as I have been told the UXG is
pretty much the same as UDM but with higher spec hardware (and fewer
services).

I appreciate that even one line changes take time, thanks for unifi-poller,
it's excellent!


Cheers,
Rob

**Comment by davidnewhall** [comment]
I'll get a new beta release out this weekend with the 1 line change. We'll see how it goes. :)

**Comment by davidnewhall** [comment]
I put 12 hours into poller today and got a lot done. I'll finish this probably next weekend.

**Comment by absoluteg** [comment]
This is AWESOME!  Thank you so much!

**Comment by light-industrial** [comment]
Woohoo!  Thanks David, much appreciated.

**Comment by davidnewhall** [comment]
The current release incorporates this feature: https://github.com/unifi-poller/unifi-poller/releases/tag/v2.1.0

Let me know if your UXG shows up and let me know what's missing. I have no way to test this myself. Please open a new issue!

**Comment by light-industrial** [comment]
Hi David,
thanks for the release, I don't see the unknown 'uxg' device type error
anymore (though can't recall if I did anything to disable that, probably
not).

However not detecting the UXG, should it show up in the "UDM/USG/" count?

unifi-poller_1  | 2021-03-22T18:21:22.329928344Z 2021/03/22 18:21:22 [INFO]
UniFi Metrics Recorded. Site: 1, Client: 34, UAP: 7, UDM/USG/: 0, USW: 4,
IDS/Event/Alarm/Anomaly: 0/0/0/0, DPI Site/Client: 0/0, Point: 94, Fields:
2691, Err: 0, Dur: 23ms
unifi-poller_1  | 2021-03-22T18:21:52.336868398Z 2021/03/22 18:21:52 [INFO]
UniFi Metrics Recorded. Site: 1, Client: 34, UAP: 7, UDM/USG/: 0, USW: 4,
IDS/Event/Alarm/Anomaly: 0/0/0/0, DPI Site/Client: 0/0, Point: 94, Fields:
2691, Err: 0, Dur: 25ms
unifi-poller_1  | 2021-03-22T18:22:22.329312042Z 2021/03/22 18:22:22 [INFO]
UniFi Metrics Recorded. Site: 1, Client: 35, UAP: 7, UDM/USG/: 0, USW: 4,
IDS/Event/Alarm/Anomaly: 0/0/0/0, DPI Site/Client: 0/0, Point: 95, Fields:
2725, Err: 0, Dur: 24ms

anything I can do to help debug?


Thanks again

**Comment by davidnewhall** [comment]
HMMMMMMM... I'm probably going to have to do some fixes. Are you running on linux, docker, mac, windows... ?

**Comment by light-industrial** [comment]
linux with docker-compose

sorry it now seems like more than just a small change

**Comment by davidnewhall** [comment]
Can you turn on debug? I'm stumped. Does debug show it unmarshalling a uxg?

**Comment by davidnewhall** [comment]
Can you try this docker tag? `golift/unifi-poller:master` Lemme know how that works out.

What to expect?
- Your UXG should appear on the USG dashboard(s).
- The application should print out a gateway counter for your UXG.
- You will find the ethernet ports on the USW dashboard(s).

**Comment by absoluteg** [comment]
@davidnewhall - I changed from Latest to what you asked above and currently get this when running "docker-compose up"

up-poller       | panic: descriptor Desc{fqName: "unifipoller_device_storage", help: "Storage", constLabels: {}, variableLabels: [type si                   te_name name source mountpoint name valtype]} is invalid: duplicate label names

Not clear on how/where I can clean this out.  I will try to do some research and see if I can resolve, but currently stuck on this failure (over and over) when trying to bring up unfi-poller.

**Comment by absoluteg** [comment]
I did try also removing docker containers and images and re-installing.  I also removed the previous container for the previous poller version i had installed (thinking might be a conflict).  None of this has worked.  I'm including the broader output when starting up:
`up-poller       | 2021/03/25 17:37:07 [INFO] Loading Configuration File: /config/unifi-poller.conf
up-poller       | 2021/03/25 17:37:07 [INFO] UniFi Poller v2.1.0 Starting Up! PID: 1
^Xup-poller       | 2021/03/25 17:37:08 [INFO] Found 1 site(s) on controller https://XXX.XXX.XXX.XXX: default (site name)
up-poller       | 2021/03/25 17:37:08 [INFO] Configured UniFi Controller 1 of 1:
up-poller       | 2021/03/25 17:37:08 [INFO]    => URL: https://xxx.xxx.xxx.xxx (verify SSL: false)
up-poller       | 2021/03/25 17:37:08 [INFO]    => Version:  (49fb554c-36de-507c-8a6c-328250f84faf)
up-poller       | 2021/03/25 17:37:08 [INFO]    => Username: user_name (has password: true)
up-poller       | 2021/03/25 17:37:08 [INFO]    => Hash PII / Poll Sites: false / all
up-poller       | 2021/03/25 17:37:08 [INFO]    => Save Sites / Save DPI: true / true (metrics)
up-poller       | 2021/03/25 17:37:08 [INFO]    => Save Events / Save IDS: false / false (logs)
up-poller       | 2021/03/25 17:37:08 [INFO]    => Save Alarms / Anomalies: false / false (logs)
up-poller       | 2021/03/25 17:37:08 [INFO]    => Save Rogue APs: false
up-poller       | 2021/03/25 17:37:08 [INFO] Poller->InfluxDB started, interval: 30s, dp: false, db: unifi, url: http://192.168.3.21:8086
up-poller       | 2021/03/25 17:37:08 [INFO] Loki config missing (or disabled), Loki output disabled!
up-poller       | 2021/03/25 17:37:08 [INFO] Internal web server disabled!
up-poller       | panic: descriptor Desc{fqName: "unifipoller_device_storage", help: "Storage", constLabels: {}, variableLabels: [type site_name name source mountpoint name valtype]} is invalid: duplicate label names
up-poller       |
up-poller       | goroutine 11 [running]:
up-poller       | github.com/prometheus/client_golang/prometheus.(*Registry).MustRegister(0xc000020aa0, 0xc00039b290, 0x1, 0x1)
up-poller       |       /go/pkg/mod/github.com/prometheus/client_golang@v1.10.0/prometheus/registry.go:403 +0xb7
up-poller       | github.com/prometheus/client_golang/prometheus.MustRegister(...)
up-poller       |       /go/pkg/mod/github.com/prometheus/client_golang@v1.10.0/prometheus/registry.go:178
up-poller       | github.com/unifi-poller/promunifi.(*promUnifi).Run(0xc0000ae000, 0xbf7fe8, 0xc0000ac020, 0x0, 0x0)
up-poller       |       /go/pkg/mod/github.com/unifi-poller/promunifi@v0.0.0-20210325083805-78fe8615bfbe/collector.go:147 +0x642
up-poller       | github.com/unifi-poller/poller.(*UnifiPoller).runOutputMethods.func1(0xc0000621e0, 0xc0000ac020, 0xc0000a6030)
up-poller       |       /go/pkg/mod/github.com/unifi-poller/poller@v0.0.0-20210325084048-71b84a5c7ed4/outputs.go:85 +0x42
up-poller       | created by github.com/unifi-poller/poller.(*UnifiPoller).runOutputMethods
up-poller       |       /go/pkg/mod/github.com/unifi-poller/poller@v0.0.0-20210325084048-71b84a5c7ed4/outputs.go:84 +0xf7
up-poller exited with code 2
`
Hoping maybe this helps and points to something easy to resolve?

**Comment by davidnewhall** [comment]
Well that's unfortunate. Please pull the `master` tag again, and try once more. Thank you!

**Comment by light-industrial** [comment]
ok debug=true, I see it unmarshalling a uxg

also, I see in the device count it mentions "UDM/USG/: 0"  .  maybe just a
logging thing, but, should there be a trailing "/UXG" on that counter?



unifi-poller_1  | 2021-03-25T18:38:32.487587855Z 2021/03/25 18:38:32.487332
updateweb.go:213: [DEBUG] Requesting
https://192.168.1.222:8443/api/stat/sites, with params: false, cookies: 2
unifi-poller_1  | 2021-03-25T18:38:32.498100837Z 2021/03/25 18:38:32.495859
updateweb.go:213: [DEBUG] Requested
https://192.168.1.222:8443/api/stat/sites: elapsed 9ms, returned 1417 bytes
unifi-poller_1  | 2021-03-25T18:38:32.498139551Z 2021/03/25 18:38:32.496230
updateweb.go:213: [DEBUG] Found 1 site(s): default
unifi-poller_1  | 2021-03-25T18:38:32.498148103Z 2021/03/25 18:38:32.496259
updateweb.go:213: [DEBUG] Polling Controller, retreiving UniFi Clients,
site default (Cloud)
unifi-poller_1  | 2021-03-25T18:38:32.498155384Z 2021/03/25 18:38:32.496327
updateweb.go:213: [DEBUG] Requesting
https://192.168.1.222:8443/api/s/default/stat/sta, with params: false,
cookies: 2
unifi-poller_1  | 2021-03-25T18:38:32.500537709Z 2021/03/25 18:38:32.500457
updateweb.go:213: [DEBUG] Requested
https://192.168.1.222:8443/api/s/default/stat/sta: elapsed 4ms, returned
46035 bytes
unifi-poller_1  | 2021-03-25T18:38:32.502563456Z 2021/03/25 18:38:32.502508
updateweb.go:213: [DEBUG] Requesting
https://192.168.1.222:8443/api/s/default/stat/device, with params: false,
cookies: 2
unifi-poller_1  | 2021-03-25T18:38:32.523152281Z 2021/03/25 18:38:32.523018
updateweb.go:213: [DEBUG] Requested
https://192.168.1.222:8443/api/s/default/stat/device: elapsed 20ms,
returned 177873 bytes
unifi-poller_1  | 2021-03-25T18:38:32.526665603Z 2021/03/25 18:38:32.526599
updateweb.go:213: [DEBUG] Unmarshalling Device Type: uap, site Cloud
(default)
unifi-poller_1  | 2021-03-25T18:38:32.528258019Z 2021/03/25 18:38:32.528183
updateweb.go:213: [DEBUG] Unmarshalling Device Type: uxg, site Cloud
(default)
unifi-poller_1  | 2021-03-25T18:38:32.531866229Z 2021/03/25 18:38:32.531802
updateweb.go:213: [DEBUG] Unmarshalling Device Type: uap, site Cloud
(default)
unifi-poller_1  | 2021-03-25T18:38:32.533689634Z 2021/03/25 18:38:32.533630
updateweb.go:213: [DEBUG] Unmarshalling Device Type: uap, site Cloud
(default)
unifi-poller_1  | 2021-03-25T18:38:32.535199597Z 2021/03/25 18:38:32.535119
updateweb.go:213: [DEBUG] Unmarshalling Device Type: usw, site Cloud
(default)
unifi-poller_1  | 2021-03-25T18:38:32.536532698Z 2021/03/25 18:38:32.536471
updateweb.go:213: [DEBUG] Unmarshalling Device Type: uap, site Cloud
(default)
unifi-poller_1  | 2021-03-25T18:38:32.538582052Z 2021/03/25 18:38:32.538521
updateweb.go:213: [DEBUG] Unmarshalling Device Type: usw, site Cloud
(default)
unifi-poller_1  | 2021-03-25T18:38:32.541180051Z 2021/03/25 18:38:32.541073
updateweb.go:213: [DEBUG] Unmarshalling Device Type: usw, site Cloud
(default)
unifi-poller_1  | 2021-03-25T18:38:32.541896936Z 2021/03/25 18:38:32.541806
updateweb.go:213: [DEBUG] Unmarshalling Device Type: uap, site Cloud
(default)
unifi-poller_1  | 2021-03-25T18:38:32.542970616Z 2021/03/25 18:38:32.542900
updateweb.go:213: [DEBUG] Unmarshalling Device Type: uap, site Cloud
(default)
unifi-poller_1  | 2021-03-25T18:38:32.545221744Z 2021/03/25 18:38:32.545120
updateweb.go:213: [DEBUG] Unmarshalling Device Type: uap, site Cloud
(default)
unifi-poller_1  | 2021-03-25T18:38:32.547294270Z 2021/03/25 18:38:32.547169
updateweb.go:213: [DEBUG] Unmarshalling Device Type: usw, site Cloud
(default)
unifi-poller_1  | 2021-03-25T18:38:32.548669938Z 2021/03/25 18:38:32.548593
updateweb.go:213: [DEBUG] Requesting
https://192.168.1.222:8443/api/stat/sites, with params: false, cookies: 2
unifi-poller_1  | 2021-03-25T18:38:32.553793597Z 2021/03/25 18:38:32.553651
updateweb.go:213: [DEBUG] Requested
https://192.168.1.222:8443/api/stat/sites: elapsed 5ms, returned 1417 bytes
unifi-poller_1  | 2021-03-25T18:38:32.553980066Z 2021/03/25 18:38:32.553936
updateweb.go:213: [DEBUG] Found 1 site(s): default
influxdb_1      | 2021-03-25T18:38:32.578085313Z [httpd] 172.19.0.1 -
unifipoller [25/Mar/2021:18:38:32 +0000] "POST
/write?consistency=&db=unifi&precision=ns&rp= HTTP/1.1" 204 0 "-"
"InfluxDBClient" <CLIENT REMOVED> 17713
unifi-poller_1  | 2021-03-25T18:38:32.578377583Z 2021/03/25 18:38:32.578248
logger.go:17: [INFO] UniFi Metrics Recorded. Site: 1, Client: 34, UAP: 7,
UDM/USG/: 0, USW: 4, IDS/Event/Alarm/Anomaly: 0/0/0/0, DPI Site/Client:
0/0, Point: 94, Fields: 2691, Err: 0, Dur: 24ms

**Comment by davidnewhall** [comment]
@light-industrial is that `master`? Doesn't look like it is.

**Comment by light-industrial** [comment]
urgh.  Apologies, I was using latest and did the pull after your release.

updated to master and, gateway detected and grafana dashboards populated,
excellent!

Thank you David!

**Comment by davidnewhall** [comment]
I'll cut another release soon with this update. For now, use `master`. If you're not using Docker, let me know what platform you're on and I'll attach a test build.

**Comment by absoluteg** [comment]
@light-industrial - are you using Docker?  I'm curious since yours is working.

**Comment by davidnewhall** [comment]
he is.

**Comment by absoluteg** [comment]
@davidnewhall - Ohhhh, no I will try as soon as I can and update.  Sorry I missed that piece.

**Comment by absoluteg** [comment]
@davidnewhall - ok I removed the docker container and image for the master pull.  I then downloaded the newest "Master" and it's running!! Yeah.  I see it listed now as "Gateways":
up-poller       | 2021/03/26 14:36:56 [INFO] UniFi Metrics Recorded. Site: 1, Client: 48, Gateways: 1, UAP: 6, USW: 8, IDS/Event/Alarm/Anomaly: 0/0/0/0, DPI Site/Client: 1/82, Point: 1090, Fields: 8121, Err: 0, Dur: 136ms

I'm now all good!!  Thank you so much!
